src/test/java/org/jenkinsci/plugins/googleplayandroidpublisher/ApkPublisherTest.java

```diff
@@ -0,0 +1,243 @@
+        // - Path or pattern to APK file was not specified
+        QueueTaskFuture<FreeStyleBuild> scheduled = p.scheduleBuild2(0);
+        j.assertBuildStatus(Result.FAILURE, scheduled);
+        j.assertLogContains("Path or pattern to APK file was not specified", scheduled.get());
+    }
+
+    @Test
+    public void uploadSingleApk_succeeds() throws Exception {
+        transport
+                .withResponse("/edits",
+                        new FakePostEditsResponse().setEditId("the-edit-id"))
+                .withResponse("/edits/the-edit-id/apks",
+                        new FakeListApksResponse().setEmptyApks())
+                .withResponse("/edits/the-edit-id/apks?uploadType=resumable",
+                        new FakeUploadApkResponse().willContinue())
+                .withResponse("google.local/uploading/foo/apk",
+                        new FakePutApkResponse().success(42, "the:sha"))
+                .withResponse("/edits/the-edit-id/tracks/production",
+                        new FakeAssignTrackResponse().success("production", 42))
+                .withResponse("/edits/the-edit-id:commit",
+                        new FakeCommitResponse().success())
+        ;
+
+        FreeStyleProject p = j.createFreeStyleProject("uploadApks");
+
+        TestsHelper.setUpCredentials("test-credentials");
+        setUpApkFile(p);
+
+        ApkPublisher publisher = new ApkPublisher();
+        publisher.setGoogleCredentialsId("test-credentials");
+        publisher.apkFilesPattern = "**/*.apk";
+        publisher.trackName = "production";
+
+        p.getPublishersList().add(publisher);
+
+        // Authenticating to Google Play API...
+        // - Credential:     test-credentials
+        // - Application ID: org.jenkins.appId
+        //
+        // Uploading 1 APK(s) with application ID: org.jenkins.appId
+        //
+        //       APK file: build/outputs/apk/app.apk
+        //     SHA-1 hash: da39a3ee5e6b4b0d3255bfef95601890afd80709
+        //    versionCode: 42
+        //  minSdkVersion: 16
+        //
+        // Setting rollout to target 100% of production track users
+        // The production release track will now contain APK(s) with version code(s): 42
+        //
+        // Applying changes to Google Play...
+        // Changes were successfully applied to Google Play
+
+        TestsHelper.assertResultWithLogLines(j, p, Result.SUCCESS,
+                "Uploading 1 APK(s) with application ID: org.jenkins.appId",
+                "APK file: " + join(Arrays.asList("build", "outputs", "apk", "app.apk"), File.separator),
+                "versionCode: 42",
+                "Setting rollout to target 100% of production track users",
+                "The production release track will now contain APK(s) with version code(s): 42",
+                "Changes were successfully applied to Google Play"
+        );
+    }
+
+    @Test
+    @WithoutJenkins
+    @Ignore("Serialization of response does not work")
```

Comment:
Will need to resolve this -- probably by only keeping the statusCode and string content in a serialized POJO, and then recreate the HttpResponse when it's needed.